### ISSUE TITLE
gtk2reactor doesn't let the UI redraw when there are is a lot of work in progress

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5067]
|<img alt="thomasvs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| thomasvs reported|
|-|-|
|Trac ID|trac#5067|
|Type|defect|
|Created|2011-04-25 15:34:29Z|
|Branch|https://github.com/twisted/twisted/tree/gtk-ui-priority-5067|

In my application I use paisley to request 15000 documents from CouchDB one by one.  Any of them can be cached already; the API has a get() which returns a deferred, by either doing a web request, or calling maybeDeferred on the cached result.

As the results from the deferred chain come in, I want the UI to update to reflect the count.

Exarkun suggested to use twisted.internet.task.cooperate, but because gtk2reactor adds timeouts with default priority (which is higher than redraw), the UI is still very unresponsive.

I made a minimal test case to compare various strategies, and the cooperate one doesn't work.

The test is at https://thomas.apestaart.org/thomas/trac/browser/tests/twisted/workui/workui.py

I'll attach the current version as well, as well as a suggested patch by exarkun that adds with lower priority and fixes this particular test.

Exarkun also mentioned [#4489](https://github.com/twisted/twisted/issues/4489) as a related ticket.

Attachments:

* [workui.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/06d/06dd5be6fbd8d391bd906e39cff1ad2c2fb1e790/aa42c733bbfdbb9149dde8d633fa21381ba2d491.py) (1593 bytes) - added by thomasvs on 2011-04-25 15:35:09Z - 
* [patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/06d/06dd5be6fbd8d391bd906e39cff1ad2c2fb1e790/d75b3b528276d7a9f30a04b8bccaf74e1d61f67a) (1116 bytes) - added by thomasvs on 2011-04-25 15:38:16Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5067 5067
type__defect defect
reporter__thomasvs thomasvs
priority__normal normal
milestone__ 
branch__branches_gtk_ui_priority_5067 branches/gtk-ui-priority-5067
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1303745669000000 1303745669000000
changetime__1332158643000000 1332158643000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

(In [33927]) Branching to 'gtk-ui-priority-5067'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [33928]) Lower priority of I/O and timeout callbacks.

Refs [#5067](https://github.com/twisted/twisted/issues/5067) and [#4489](https://github.com/twisted/twisted/issues/4489)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set status to `closed`|
|-|-|

(In [33931]) Merge gtk-ui-priority-5067: Lower priority for Twisted callbacks in gtk/glib.

Author: thomasvs, exarkun
Review: itamar
Fixes: [#4489](https://github.com/twisted/twisted/issues/4489), [#5067](https://github.com/twisted/twisted/issues/5067)

Lower the priority of Twisted events in the glib/gtk reactor so that the UI will always refresh.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

I suppose Gtk integration is one area where we will relax our unit testing requirements, since Gtk itself isn't particularly friendly to unit testing this sort of thing?

Just posting this comment to make it explicit that this is why this change landed with no unit tests, which would normally be in contradiction to our development process.

I'll also take the review points, since they're still sitting there unclaimed. ;)
